### PR TITLE
Making yaml tests version selector parser compatible with versions returned by Build

### PR DESCRIPTION
**Pull request by ldematte**
Spinoff from https://github.com/elastic/elasticsearch/pull/100770
- Makes version matching work even when it includes `-SNAPSHOT` (we provisionally have mixed use of Build.version() and Version.toString() for the Rest APIs)
- Makes `version: current` work with non-semantic versions, moving it to a string.equals comparison

### CODE REVIEW

**Comment by ldematte** [comment]
@elasticsearchmachine rerun elasticsearch-ci/bwc

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java [hunk 1]
```diff
@@ -626,32 +628,55 @@ public String toString() {
...
-        List<VersionRange> skipVersionRanges = parser.text().equals("current")
-            ? List.of(new VersionRange(Version.CURRENT, Version.CURRENT))
-            : SkipSection.parseVersionRanges(parser.text());
+
+        final Predicate<String> nodeMatcher;
+        final String versionSelectorString;
+        if (parser.text().equals("current")) {
+            var currentUnqualified = Build.current().version().replace("-SNAPSHOT", "");
+            nodeMatcher = nodeVersion -> currentUnqualified.equals(nodeVersion.replace("-SNAPSHOT", ""));
+            versionSelectorString = "current";
```
Comment (rjernst) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java:655]: Should this be set to `currentUnqualified` (or perhaps a string containing both, to make it clear that current was set, but that it was replaced with the actual current version)?

File: test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/DoSectionTests.java [hunk 2]
```diff
@@ -626,6 +629,28 @@ public void testNodeSelectorByVersion() throws IOException {
         }
     }
 
+    public void testNodeSelectorByVersionRangeFailsWithNonSemanticVersion() throws IOException {
```
Comment (rjernst) [inline test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/DoSectionTests.java:632]: How about a test of "current"?

**Review by rjernst — COMMENTED** [review]
It looks like part of this PR is adding a new "current" version specifier. What's the case that is needed?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java (hunk 1 above)
Comment (ldematte) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java:655]: Good suggestion

**Comment by ldematte** [comment]
> It looks like part of this PR is adding a new "current" version specifier. What's the case that is needed?

"current" was already there. The difference is that it worked by comparing with a (Version.CURRENT, Version.CURRENT) range. For all purposes this should be the same, and in fact the existing tests with current are still passing.

File: test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/DoSectionTests.java (hunk 2 above)
Comment (ldematte) [reply test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/DoSectionTests.java:632]: "current" was already there, I simply changed it in a way that is compatible but does not use Version.CURRENT; a test `testNodeSelectorCurrentVersion` was already there (and it's still passing).

Comment (rjernst) [reply test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/DoSectionTests.java:632]: Ah thanks, couldn't see it without expanding :)

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**